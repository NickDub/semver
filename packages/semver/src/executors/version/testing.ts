import { promises } from 'fs';
import { resolve } from 'path';
import * as tmp from 'tmp';
import {
  ExecutorContext,
  ProjectConfiguration,
  TargetConfiguration,
} from '@nx/devkit';
import { readJsonFile, workspaceRoot, writeJsonFile } from '@nx/devkit';
import { execSync } from 'child_process';

export interface TestingWorkspace {
  exec(command: string): void;
  runNx(command: string): void;
  generateLib(name: string, options?: string): void;
  installSemver(project: string, preset?: string): void;
  tearDown(): Promise<void>;
  root: string;
}

const packageManager = 'npm';

function runNxNewCommand(dir: string) {
  execSync(
    `node ${require.resolve(
      'nx',
    )} new proj --nx-workspace-root=${dir} --no-interactive --skip-install --collection=@nx/workspace --npmScope=proj --preset=apps --package-manager=${packageManager}`,
    {
      cwd: dir,
      stdio: 'ignore',
    },
  );
}

function linkPackage(dir: string) {
  const json = readJsonFile(resolve(dir, 'package.json'));
  json.devDependencies = {
    ...json.devDependencies,
    '@jscutlery/semver': `file:${resolve(
      workspaceRoot,
      'dist/packages/semver',
    )}`,
  };
  writeJsonFile(resolve(dir, 'package.json'), json);
}

function runInstall(dir: string) {
  execSync(`${packageManager} install`, {
    cwd: dir,
    stdio: 'ignore',
  });
}

function initGit(dir: string) {
  execSync(
    `
        git init --quiet

        # These are needed by CI.
        git config user.email "bot@jest.io"
        git config user.name "Test Bot"
        git config commit.gpgsign false
`,
    { cwd: dir, stdio: 'ignore' },
  );
}

export function setupTestingWorkspace(): TestingWorkspace {
  /* Create a temporary directory. */
  const tmpDir = tmp.dirSync();

  const originalCwd = process.cwd();
  process.chdir(tmpDir.name);

  const tmpRoot = process.cwd();
  const workspaceRoot = resolve(tmpRoot, 'proj');

  runNxNewCommand(tmpRoot);
  initGit(workspaceRoot);
  linkPackage(workspaceRoot);
  runInstall(workspaceRoot);

  return {
    runNx(command: string) {
      execSync(`node ${require.resolve('nx')} ${command}`, {
        cwd: workspaceRoot,
        stdio: 'inherit',
      });
    },

    exec(command: string) {
      execSync(command, {
        cwd: workspaceRoot,
        stdio: 'inherit',
      });
    },

    generateLib(name: string, options = '') {
      const commonArgs = `--directory=libs --unitTestRunner=none --linter=none --bundler=none --minimal --skipFormat`;
      this.runNx(`g @nx/js:lib ${name} ${commonArgs} ${options}`);
    },

    installSemver(project: string, options = '') {
      this.runNx(
        `g @jscutlery/semver:install --projects=${project} ${options}`,
      );
    },

    async tearDown() {
      await promises.rm(tmpRoot, { recursive: true });
      process.chdir(originalCwd);
    },

    root: workspaceRoot,
  };
}

export function createFakeContext({
  cwd = process.cwd(),
  project,
  projectRoot,
  workspaceRoot,
  targets = {},
  additionalProjects = [],
}: {
  cwd?: string;
  project: string;
  projectRoot: string;
  workspaceRoot: string;
  targets?: Record<string, TargetConfiguration>;
  additionalProjects?: {
    project: string;
    projectRoot: string;
    targets?: Record<string, TargetConfiguration>;
  }[];
}): ExecutorContext {
  return {
    isVerbose: false,
    cwd: cwd,
    root: workspaceRoot,
    projectName: project,
    projectsConfigurations: {
      version: 2,
      projects: {
        [project]: {
          root: projectRoot,
          targets,
        },
        ...assembleAdditionalProjects(additionalProjects),
      },
    },
    projectGraph: {
      nodes: {},
      dependencies: {},
    },
  } satisfies ExecutorContext;
}

function assembleAdditionalProjects(
  additionalProjects: {
    project: string;
    projectRoot: string;
    targets?: Record<string, TargetConfiguration>;
  }[],
) {
  return additionalProjects.reduce<{
    [projectName: string]: ProjectConfiguration;
  }>(
    (acc, p) => {
      acc[p.project] = {
        root: p.projectRoot,
        targets: p.targets || {},
      };
      return acc;
    },
    {} satisfies { [project: string]: ProjectConfiguration },
  );
}
