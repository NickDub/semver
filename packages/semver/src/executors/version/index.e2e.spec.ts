import { execSync } from 'child_process';
import { setupTestingWorkspace, type TestingWorkspace } from './testing';
import { readFileSync } from 'fs';

describe('@jscutlery/semver', () => {
  let testingWorkspace: TestingWorkspace;

  describe('package "a"', () => {
    beforeAll(async () => {
      testingWorkspace = setupTestingWorkspace();
      testingWorkspace.runNx(
        `g @nx/js:lib a --directory=libs --unitTestRunner=none --linter=none --bundler=none --minimal --publishable --importPath=@proj/a`,
      );
      testingWorkspace.exec(
        `
          git add .
          git commit -m "🐣"
        `,
      );
      testingWorkspace.runNx(`g @jscutlery/semver:install --projects=a`);
      testingWorkspace.exec(
        `
          git add .
          git commit -m "build: 📦 setup semver"
        `,
      );
      testingWorkspace.exec(
        `
          echo feat > libs/a/a.txt
          git add .
          git commit -m "feat(a): 🚀 new feature"

          echo fix >> libs/a/a.txt
          git add .
          git commit -m "fix(a): 🐞 fix bug"
        `,
      );
      // @TODO: Remove --noVerify when "release" commit type is allowed by commitlint.
      testingWorkspace.runNx(`run a:version --noVerify`);
    });

    afterAll(() => testingWorkspace.tearDown());

    it('should commit all changes', () => {
      expect(uncommitedChanges(testingWorkspace.root)).toHaveLength(0);
    });

    it('should tag with version', () => {
      expect(getLastTag(testingWorkspace.root)).toBe('a-0.1.0');
    });

    it('should create only one tag', () => {
      expect(getTags(testingWorkspace.root)).toHaveLength(1);
    });

    it('should commit with description', () => {
      expect(getLastCommitDescription(testingWorkspace.root)).toBe(
        'chore(a): release version 0.1.0',
      );
    });

    it('should bump package version', () => {
      expect(readFile(`${testingWorkspace.root}/libs/a/package.json`)).toMatch(
        /"version": "0.1.0"/,
      );
    });

    it('should generate CHANGELOG.md', () => {
      expect(readFile(`${testingWorkspace.root}/libs/a/CHANGELOG.md`)).toMatch(
        new RegExp(`^# Changelog

This file was generated.*

# 0.1.0 \\(.*\\)


### Bug Fixes

\\* \\*\\*a:\\*\\* 🐞 fix bug .*


### Features

\\* \\*\\*a:\\*\\* 🚀 new feature .*
$`),
      );
    });
  });
});

function getLastTag(dir: string) {
  return execSync('git describe --tags --abbrev=0', {
    encoding: 'utf-8',
    cwd: dir,
    stdio: 'ignore',
  }).trim();
}

function getLastCommitDescription(dir: string) {
  return execSync('git log -1 --pretty=%B', {
    encoding: 'utf-8',
    cwd: dir,
    stdio: 'ignore',
  }).trim();
}

function getTags(dir: string) {
  return execSync('git tag', {
    encoding: 'utf-8',
    cwd: dir,
    stdio: 'ignore',
  })
    .trim()
    .split('\n');
}

function readFile(path: string) {
  return readFileSync(path, {
    encoding: 'utf-8',
  });
}

function uncommitedChanges(dir: string) {
  return (
    execSync('git status --porcelain', {
      encoding: 'utf-8',
      cwd: dir,
    })
      .split('\n')
      /* Remove empty line. */
      .filter((line) => line.length !== 0)
  );
}
