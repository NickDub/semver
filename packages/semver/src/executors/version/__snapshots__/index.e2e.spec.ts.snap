// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@jscutlery/semver @jscutlery/semver:version when libs/a changed (breaking change) should generate CHANGELOG.md 1`] = `
"# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

# [1.0.0](/compare/a-0.1.0...a-1.0.0) (yyyy-mm-dd)


### Features

* **a:** 🚀 new feature xxxxxxx


### BREAKING CHANGES

* **a:** 🚨 Breaking change description



# 0.1.0 (yyyy-mm-dd)


### Bug Fixes

* **a:** 🐞 fix bug xxxxxxx


### Features

* **a:** 🚀 new feature xxxxxxx
"
`;

exports[`@jscutlery/semver @jscutlery/semver:version when libs/a changed should generate CHANGELOG.md 1`] = `
"# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

# 0.1.0 (yyyy-mm-dd)


### Bug Fixes

* **a:** 🐞 fix bug xxxxxxx


### Features

* **a:** 🚀 new feature xxxxxxx
"
`;

exports[`@jscutlery/semver @jscutlery/semver:version when libs/b changed (with --skipCommit) should generate CHANGELOG.md 1`] = `
"# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

## [0.2.0](///compare/b-0.1.0...b-0.2.0) (yyyy-mm-dd)


### Features

* **b:** 🚀 new feature xxxxxxx

## 0.1.0 (yyyy-mm-dd)


### Features

* **b:** 🚀 new feature xxxxxxx


### Bug Fixes

* **b:** 🐞 fix bug xxxxxxx
"
`;

exports[`@jscutlery/semver @jscutlery/semver:version when libs/b changed should generate CHANGELOG.md 1`] = `
"# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

## 0.1.0 (yyyy-mm-dd)


### Features

* **b:** 🚀 new feature xxxxxxx


### Bug Fixes

* **b:** 🐞 fix bug xxxxxxx
"
`;

exports[`@jscutlery/semver @jscutlery/semver:version when libs/d changed should generate CHANGELOG.md 1`] = `
"# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

## 0.0.0 (yyyy-mm-dd)


### Awesome features

* **d:** 🚀 new awesome feature xxxxxxx
"
`;
